static/app/views/issueList/utils/issueViews.tsx

```diff
@@ -0,0 +1,556 @@
+import {useNavigate} from 'sentry/utils/useNavigate';
+import useOrganization from 'sentry/utils/useOrganization';
+import usePageFilters from 'sentry/utils/usePageFilters';
+import {generateTempViewId} from 'sentry/views/issueList/groupSearchViewTabs/draggableTabBar';
+import {useUpdateGroupSearchViews} from 'sentry/views/issueList/mutations/useUpdateGroupSearchViews';
+import type {
+  GroupSearchView,
+  UpdateGroupSearchViewPayload,
+} from 'sentry/views/issueList/types';
+import {IssueSortOptions} from 'sentry/views/issueList/utils';
+
+const TEMPORARY_TAB_KEY = 'temporary-tab';
+
+export interface IssueView {
+  id: string;
+  /**
+   * False for tabs that were added view the "Add View" button, but
+   * have not been edited in any way. Only tabs with isCommitted=true
+   * will be saved to the backend.
+   */
+  isCommitted: boolean;
+  key: string;
+  label: string;
+  query: string;
+  querySort: IssueSortOptions;
+  content?: React.ReactNode;
+  unsavedChanges?: [string, IssueSortOptions];
+}
+
+type ReorderTabsAction = {
+  newKeyOrder: string[];
+  type: 'REORDER_TABS';
+};
+
+type SaveChangesAction = {
+  type: 'SAVE_CHANGES';
+};
+
+type DiscardChangesAction = {
+  type: 'DISCARD_CHANGES';
+};
+
+type RenameTabAction = {
+  newLabel: string;
+  type: 'RENAME_TAB';
+};
+
+type DuplicateViewAction = {
+  newViewId: string;
+  type: 'DUPLICATE_VIEW';
+};
+
+type DeleteViewAction = {
+  type: 'DELETE_VIEW';
+};
+
+type CreateNewViewAction = {
+  tempId: string;
+  type: 'CREATE_NEW_VIEW';
+};
+
+type SetTempViewAction = {
+  query: string;
+  sort: IssueSortOptions;
+  type: 'SET_TEMP_VIEW';
+};
+
+type DiscardTempViewAction = {
+  type: 'DISCARD_TEMP_VIEW';
+};
+
+type SaveTempViewAction = {
+  type: 'SAVE_TEMP_VIEW';
+};
+
+type UpdateUnsavedChangesAction = {
+  type: 'UPDATE_UNSAVED_CHANGES';
+  unsavedChanges: [string, IssueSortOptions] | undefined;
+  isCommitted?: boolean;
+  syncViews?: boolean;
+};
+
+type UpdateViewIdsAction = {
+  newViews: UpdateGroupSearchViewPayload[];
+  type: 'UPDATE_VIEW_IDS';
+};
+
+type SetViewsAction = {
+  type: 'SET_VIEWS';
+  views: IssueView[];
+  syncViews?: boolean;
+};
+
+type SyncViewsToBackendAction = {
+  type: 'SYNC_VIEWS_TO_BACKEND';
+};
+
+export type IssueViewsActions =
+  | ReorderTabsAction
+  | SaveChangesAction
+  | DiscardChangesAction
+  | RenameTabAction
+  | DuplicateViewAction
+  | DeleteViewAction
+  | CreateNewViewAction
+  | SetTempViewAction
+  | DiscardTempViewAction
+  | SaveTempViewAction
+  | UpdateUnsavedChangesAction
+  | UpdateViewIdsAction
+  | SetViewsAction
+  | SyncViewsToBackendAction;
+
+export interface IssueViewsState {
+  views: IssueView[];
+  tempView?: IssueView;
+}
+
+export interface IssueViewsContextType extends TabContext {
+  dispatch: Dispatch<IssueViewsActions>;
+  state: IssueViewsState;
+}
+
+export const IssueViewsContext = createContext<IssueViewsContextType>({
+  rootProps: {orientation: 'horizontal'},
```

Comment:
Is this necessary to be in the context?

Reply:
This is one of react-aria's tab context variables that I included since this context is essentially just a wrapper on that one. I copied the behavior from our base tabs component [here](https://github.com/getsentry/sentry/blob/e8371a7e81a05fdae24eec4d5ed042a536697cbf/static/app/components/tabs/index.tsx#L70).